File: tbm4linux.py
#! /usr/bin/python3

import argparse
import json
import os
import re
import shutil

import requests

BUCKET_PATH = "bucket"
CACHE_PATH = "/tmp"
BINARY_PATH = os.path.expanduser("~/.local/bin")
FOLDER_PATH = os.path.expanduser("~/.local")
EXTRACT_SCRIPT_PATH = "/usr/local/bin/extract.sh"

ARCHITECTURE = "x64"


def read_config(config_file):
    with open(config_file) as file:
        config = json.loads(file.read())
    version = config["version"]
    formatted_config = json.loads(json.dumps(config).replace("<version>", version))
    return config, formatted_config


def update_config(config, config_file):
    config_str = json.dumps(config, indent=4)
    with open(config_file, "w") as file:
        file.write(config_str)


def check_version(checkver_dict):
    resp = requests.get(checkver_dict["url"], headers={"user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0"})
    html = resp.text
    new_version = re.search(checkver_dict["pattern"], html).group(1)
    return new_version


def install(config_file):
    _, formatted_config = read_config(config_file)
    asset_name = formatted_config["architecture"][ARCHITECTURE]["asset_name"]
    if not os.path.exists(f"{CACHE_PATH}/{id}"):
        os.mkdir(f"{CACHE_PATH}/{id}")
    os.chdir(f"{CACHE_PATH}/{id}")

    print("\tdownloading...", end="")
    file_response = requests.get(formatted_config["architecture"][ARCHITECTURE]["url"])
    with open(asset_name, "wb") as file:
        file.write(file_response.content)
    print("ok", end="")

    if formatted_config["architecture"][ARCHITECTURE]["extract"]:
        print("...extracting...", end="")
        try:
            os.system(f"{EXTRACT_SCRIPT_PATH} {asset_name}")
            print("ok")
        except:
            print("failed!")
            return
    else:
        print()
    for k, v in formatted_config["architecture"][ARCHITECTURE]["folder"].items():
        dst = os.path.join(FOLDER_PATH, v)
        print(f"\t{k} -> {dst}")
        if os.path.exists(dst):
            if input(f"\tconfirm that {dst} is going to be deleted(y/n): ") == "y":
                shutil.rmtree(dst)
            else:
                print(f"\t{dst} not updated")
                continue
        shutil.copytree(k, dst)
    for k, v in formatted_config["architecture"][ARCHITECTURE]["bin"].items():
        dst = os.path.join(BINARY_PATH, v)
        print(f"\t{k} -> {dst}")
        shutil.copy2(k, dst)
        os.chmod(dst, 0o755)
    os.chdir(cwd)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Linux Binary Manager")
    parser.add_argument("ids", nargs="+", help="binary id")
    parser.add_argument("--check", "-c", action="store_true", help="Check for updates")
    parser.add_argument("--install", "-i", action="store_true", help="Install the binary")
    parser.add_argument("--update", "-u", action="store_true", help="Check for updates and install")
    args = parser.parse_args()

    if args.ids[0] == "*":
        ids = [file.split(".")[0] for file in os.listdir(BUCKET_PATH) if file.endswith(".json")]
    else:
        ids = args.ids

    cwd = os.getcwd()
    for idx, id in enumerate(ids):
        print(f"({idx+1:03}/{len(ids):03})[{id}]: ", end="")
        config_file = f"{BUCKET_PATH}/{id}.json"
        if os.path.exists(config_file):
            if args.check or args.update:
                config, formatted_config = read_config(config_file)
                old_version = config["version"]
                new_version = check_version(config["checkver"])
                if new_version != old_version:
                    print(f"<{old_version}> -> <{new_version}>")
                    config["version"] = new_version
                    update_config(config, config_file)
                    if args.update:
                        install(config_file)
                else:
                    print(f"<{old_version}>")
            if args.install:
                install(config_file)
        else:
            print(f"config file {config_file} not found")
        print(f"({idx+1:03}/{len(ids):03})[{id}]: done")

File: tbm4linux.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
)

type Config struct {
	ExampleURL   string `json:"example_url"`
	Version      string `json:"version"`
	Architecture map[string]struct {
		Url       string            `json:"url"`
		AssetName string            `json:"asset_name"`
		Extract   bool              `json:"extract"`
		Bin       map[string]string `json:"bin"`
		Folder    map[string]string `json:"folder"`
	} `json:"architecture"`
	Checkver map[string]string `json:"checkver"`
}

var (
	bucketPath        = "bucket"
	cachePath         = "/tmp"
	homeDir, _        = os.UserHomeDir()
	binanryPath       = filepath.Join(homeDir, ".local/bin")
	folderPath        = filepath.Join(homeDir, ".local")
	extractScriptPath = "/usr/local/bin/extract.sh"

	arch = "x64"
)

const (
	CONCURRENCY = 10
)

func main() {
	checkFlag := flag.Bool("c", false, "Check for updates")
	installFlag := flag.Bool("i", false, "Install the binary")
	updateFlag := flag.Bool("u", false, "Check for updates and install")
	flag.Parse()
	args := flag.Args()

	var ids []string
	if len(args) == 0 {
		fmt.Println("Please provide at least one binary id")
		return
	}

	if args[0] == "*" {
		files, err := filepath.Glob(filepath.Join(bucketPath, "*.json"))
		if err != nil {
			fmt.Println("Error:", err)
			return
		}
		for _, file := range files {
			id := strings.TrimSuffix(filepath.Base(file), ".json")
			ids = append(ids, id)
		}
	} else {
		ids = args[:]
	}
	var final []string
	if *checkFlag || *updateFlag {
		sem := make(chan bool, CONCURRENCY)
		var mutex sync.Mutex
		for _, id := range ids {
			sem <- true
			go func(id string) {
				defer func() { <-sem }()
				configFile := filepath.Join(bucketPath, id+".json")
				if _, err := os.Stat(configFile); err == nil {
					config, formattedConfig := readConfig(configFile)
					oldVersion := config.Version
					newVersion := checkVersion(formattedConfig)
					if newVersion != oldVersion {
						fmt.Printf("[%s]: <%s> -> <%s>\n", id, oldVersion, newVersion)
						config.Version = newVersion
						updateConfig(config, configFile)
						mutex.Lock()
						final = append(final, id)
						mutex.Unlock()
					} else {
						fmt.Printf("[%s]: <%s>\n", id, oldVersion)
					}
				}
			}(id)
		}
		for i := 0; i < cap(sem); i++ {
			sem <- true
		}
	} else {
		final = ids[:]
	}

	if *updateFlag || *installFlag {
		fmt.Println("********Installing********")
		for _, id := range final {
			configFile := filepath.Join(bucketPath, id+".json")
			err := install(id, configFile)
			if err != nil {
				fmt.Printf("[%s]: error installing binary. %v\n", id, err)
				continue
			}
			fmt.Printf("[%s]: done\n", id)
		}
	}
}

func readConfig(configFile string) (config Config, formattedConfig Config) {
	content, _ := os.ReadFile(configFile)
	json.Unmarshal(content, &config)
	formattedContent := []byte(strings.ReplaceAll(string(content), "<version>", config.Version))
	json.Unmarshal(formattedContent, &formattedConfig)
	return config, formattedConfig
}

func updateConfig(config Config, configFile string) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	enc.SetIndent("", "    ")
	enc.Encode(config)
	os.WriteFile(configFile, buf.Bytes(), 0644)
}

func checkVersion(config Config) string {
	req, _ := http.NewRequest("GET", config.Checkver["url"], nil)
	req.Header.Add("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:124.0) Gecko/20100101 Firefox/124.0")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return ""
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	html := string(body)
	re := regexp.MustCompile(config.Checkver["pattern"])
	match := re.FindStringSubmatch(html)
	if len(match) < 2 {
		return ""
	}
	newVersion := match[1]
	return newVersion
}

func install(id string, configFile string) error {
	_, formattedConfig := readConfig(configFile)
	assetName := formattedConfig.Architecture[arch].AssetName
	workDir := filepath.Join(cachePath, id)

	if _, err := os.Stat(workDir); os.IsNotExist(err) {
		os.Mkdir(workDir, 0755)
	}
	cwd, _ := os.Getwd()
	os.Chdir(workDir)
	defer os.Chdir(cwd)
	fmt.Printf("[%s]: downloading...", id)
	fileResponse, err := http.Get(formattedConfig.Architecture[arch].Url)
	if err != nil {
		return err
	}
	defer fileResponse.Body.Close()
	fileContent, _ := io.ReadAll(fileResponse.Body)
	os.WriteFile(filepath.Join(workDir, assetName), fileContent, 0644)
	fmt.Print("ok")

	if formattedConfig.Architecture[arch].Extract {
		fmt.Print("...extracting...")
		err := exec.Command(extractScriptPath, assetName).Run()
		if err != nil {
			fmt.Println("failed!")
			return err
		}
		fmt.Println("ok")
	} else {
		fmt.Println()
	}

	for k, v := range formattedConfig.Architecture[arch].Folder {
		dst := filepath.Join(folderPath, v)
		fmt.Printf("[%s]: %s -> %s\n", id, k, dst)
		if _, err := os.Stat(dst); err == nil {
			fmt.Printf("[%s]: confirm that %s is going to be deleted(y/n): ", id, dst)
			var confirmation string
			fmt.Scanln(&confirmation)
			if strings.ToLower(confirmation) == "y" {
				os.RemoveAll(dst)
			} else {
				fmt.Printf("[%s]: %s not updated\n", id, dst)
				continue
			}
		}
		exec.Command("cp", "-r", k, dst).Run()

	}
	for k, v := range formattedConfig.Architecture[arch].Bin {
		dst := filepath.Join(binanryPath, v)
		fmt.Printf("[%s]: %s -> %s\n", id, k, dst)
		exec.Command("install", "-m", "0755", k, dst).Run()

	}
	return nil
}
